### PR TITLE
Diagnostic namespace

### PR DESCRIPTION
**Pull request by weiznich**
This PR implements the basic infrastructure for accepting the `#[diagnostic]` attribute tool namespace as specified in https://github.com/rust-lang/rfcs/pull/3368. Note: This RFC is not merged yet, but it seems like it will be accepted soon. I open this PR early on to get feedback on the actual implementation as soon as possible. This hopefully enables getting at least the diagnostic namespace to stable rust "soon", so that crates do not need to bump their MSRV if we stabilize actual attributes in this namespace.

 This PR only adds infrastructure accept attributes from this namespace, it does not add any specific attribute. Therefore the compiler will emit a lint warning for each attribute that's actually used. This namespace is added behind a feature flag, so it will be only available on a nightly compiler for now.

cc @estebank as they've supported me in planing, specifying and implementing this feature.

### CODE REVIEW

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -381,6 +381,8 @@ declare_features! (
     (active, deprecated_safe, "1.61.0", Some(94978), None),
     /// Allows having using `suggestion` in the `#[deprecated]` attribute.
     (active, deprecated_suggestion, "1.61.0", Some(94785), None),
+    /// Allows using the `#[diagnostic]` attribute tool namespace
+    (active, diagnostic_namespace, "CURRENT_RUSTC_VERSION", Some(94785), None),
```
Comment (weiznich) [inline compiler/rustc_feature/src/active.rs:385]: TODO: Add a tracking issue and put the actual issue number in here.

**Comment by matthewjasper** [comment]
cc @petrochenkov

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 2]
```diff
@@ -4175,3 +4176,27 @@ declare_lint! {
...
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to specify a diagnostic attribute that does not exist. Check
+    /// the spelling, and check the diagnostic attribute listing for the correct name. Also
+    /// consider if you are using an old version of the compiler, and the attribute
+    /// is only available in a newer version.
+    pub UNKNOWN_DIAGNOSTIC_ATTRIBUTE,
```
Comment (est31) [inline compiler/rustc_lint_defs/src/builtin.rs:4199]: This should be plural as per [RFC 344](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints).

I also wonder what to do about diagnostic attributes that use a different and unknown format. E.g. previously one would only be allowed to specify `#[diagnostic::on_unimplemented = "abc def"]`, but now one can do `#[diagnostic::on_unimplemented(note = "hello", help = "some help note here")]`, or even `#[diagnostic::on_unimplemented(fluent_slug = "some_name")]`. What lint would then be fired, and if it is this lint, which is fine, is the name good?

Comment (weiznich) [reply compiler/rustc_lint_defs/src/builtin.rs]: I've fixed the lint name in https://github.com/rust-lang/rust/pull/111780/commits/4f20a4f3f95355f692fa30c406d0d6914935b3b5

As for the unknown options variant: I would expect that we also implement a (different) lint there. At least as crate author I might want to allow new/changed options but no unknown attributes. This other lint can only be implemented after there is an actual attribute.

Comment (est31) [reply compiler/rustc_lint_defs/src/builtin.rs]: @weiznich thanks! The proposed strategy to have two lints sounds good.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 3]
```diff
@@ -4175,3 +4176,27 @@ declare_lint! {
     Warn,
     "\"invalid_parameter\" isn't a valid argument for `#[macro_export]`",
 }
+
+declare_lint! {
+    /// The `unknown_diagnostic_attribute` lint detects unrecognized diagnostic attributes.
```
Comment (est31) [inline compiler/rustc_lint_defs/src/builtin.rs:4181]: missed this one :)

```suggestion
    /// The `unknown_diagnostic_attributes` lint detects unrecognized diagnostic attributes.
```

File: compiler/rustc_feature/src/active.rs (hunk 1 above)
Comment (weiznich) [reply compiler/rustc_feature/src/active.rs:385]:
```suggestion
    (active, diagnostic_namespace, "CURRENT_RUSTC_VERSION", Some(111996), None),
```

File: compiler/rustc_hir/src/def.rs [hunk 4]
```diff
@@ -405,7 +405,7 @@ pub enum Res<Id = hir::HirId> {
     /// A tool attribute module; e.g., the `rustfmt` in `#[rustfmt::skip]`.
     ///
     /// **Belongs to the type namespace.**
-    ToolMod,
+    ToolMod { allow_shadow: bool },
```
Comment (petrochenkov) [inline compiler/rustc_hir/src/def.rs:408]:
```suggestion
    ToolMod,
```
This feature should not introduce special ad hoc rules on shadowing like this.
I specifically asked on the RFC whether `diagnostic::foo` should be tool attributes and the answer was yes, so I don't think it should additionally bend language rules to serve its specific case, it already barely pulls its weight.

Comment (weiznich) [reply compiler/rustc_hir/src/def.rs:408]: This is somewhat required to workaround https://github.com/rust-lang/rust/issues/111135

To cite from the [tool attribute RFC](https://rust-lang.github.io/rfcs/2103-tool-attributes.html):

> During macro expansion, when faced with an attribute, the compiler first tries to find a macro using the macro name resolution rules. The compiler then checks if the attribute matches any of the declared or built- in attributes. 

Which means at least as far as I understand it's totally fine to allow shadowing there. 

I've chosen to make this shadowing behaviour optional and use it only for the new namespace for now to minimize possible breakage there. If you feel that's the wrong solution please suggest a better way to fix the underlying bug.

> I don't think it should additionally bend language rules to serve its specific case, it already barely pulls its weight.

I do not think that's the correct place to discuss whether or not the namespace pulls in its weight. It seems like that the language team feels that it pulls in it's weight, so you might want to raise that concern there.

Comment (petrochenkov) [reply compiler/rustc_hir/src/def.rs:408]: >The compiler then checks if the attribute matches any of the declared or built- in attributes.

Then this part is the random ad hoc language bending that should not be in the RFC.

>If you feel that's the wrong solution please suggest a better way to fix the underlying bug.

I think the better way is to remove the artificial errors from https://github.com/rust-lang/rust/issues/56414.
That issue says "this restriction can be lifted fully or partially if it causes too much trouble in practice with ambiguity errors, or with addition of new names to libraries being backward-incompatible due to new ambiguities", and it did consistently caused at least some amount of trouble, with `diagnostic` being just a single instance of that.

Comment (weiznich) [reply compiler/rustc_hir/src/def.rs:408]: > Then this part is the random ad hoc language bending that should not be in the RFC.

I don't think it's worth to argue about that. It is accepted stable language behaviour since 2016 according to RFC-2103 (tool attributes).

> I think the better way is to remove the artificial errors from #56414.
That issue says "this restriction can be lifted fully or partially if it causes too much trouble in practice with ambiguity errors, or with addition of new names to libraries being backward-incompatible due to new ambiguities", and it did consistently caused at least some amount of trouble, with diagnostic being just a single instance of that.

Just to be sure that I understand that correctly: You basically suggest to remove the `allow_shadow` flag and to suppress unconditionally the ambiguity error [here](https://github.com/rust-lang/rust/pull/111780/files#diff-c046507afdba3b0705638f53fffa156cbad72ed17aa01d96d7bd1cc10b8d9bceR639)?

Comment (petrochenkov) [reply compiler/rustc_hir/src/def.rs:408]: >I don't think it's worth to argue about that. It is accepted stable language behaviour since 2016 according to RFC-2103 (tool attributes).

Ah, that's from the tool attribute RFC.
It's not the actually implemented stable language behavior.
RFC details are typically tweaked based on implementation experience, and in this specific case the RFC text is especially outdated because it pre-dates macro modularization and uniform imports.
My suggestion is to follow the rules that are actually implemented since 2018 edition.

>Just to be sure that I understand that correctly: You basically suggest to remove the allow_shadow flag and to suppress unconditionally the ambiguity error [here](https://github.com/rust-lang/rust/pull/111780/files#diff-c046507afdba3b0705638f53fffa156cbad72ed17aa01d96d7bd1cc10b8d9bceR639)?

I need to check what exactly needs to be removed to fix https://github.com/rust-lang/rust/issues/56414, probably on Monday.

Comment (petrochenkov) [reply compiler/rustc_hir/src/def.rs:408]: I implemented the fix for https://github.com/rust-lang/rust/issues/56414 in https://github.com/rust-lang/rust/pull/112086.

Comment (weiznich) [reply compiler/rustc_hir/src/def.rs:408]: Thanks for putting up a fix that fast. i will rebase my PR on top of yours and remove the special handling in the next few days.

**Comment by weiznich** [comment]
I've rebased this on top of https://github.com/rust-lang/rust/pull/112086 to resolve @petrochenkov concerns. Merging this PR needs to wait till #112086 has landed.

File: compiler/rustc_hir/src/def.rs [hunk 5]
```diff
@@ -630,7 +630,7 @@ impl<Id> Res<Id> {
             | Res::SelfTyParam { .. }
             | Res::SelfTyAlias { .. }
             | Res::SelfCtor(..)
-            | Res::ToolMod
+            | Res::ToolMod { .. }
```
Comment (estebank) [inline compiler/rustc_hir/src/def.rs:633]: I believe that all the `Res::ToolMod { .. }` changes can be reverted to not have the braces (keeping this PR's diff shorter).

Comment (weiznich) [reply compiler/rustc_hir/src/def.rs]: Fixed in https://github.com/rust-lang/rust/pull/111780/commits/08744dc43d48a2e70369e33af49cd4144a81a599

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 6]
```diff
@@ -1520,7 +1520,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         if b.span.is_dummy() || !self.tcx.sess.source_map().is_span_accessible(b.span) {
             // These already contain the "built-in" prefix or look bad with it.
             let add_built_in =
-                !matches!(b.res(), Res::NonMacroAttr(..) | Res::PrimTy(..) | Res::ToolMod);
+                !matches!(b.res(), Res::NonMacroAttr(..) | Res::PrimTy(..) | Res::ToolMod { .. });
```
Comment (estebank) [inline compiler/rustc_resolve/src/diagnostics.rs:1523]: (This change is no longer needed)

File: compiler/rustc_resolve/src/macros.rs [hunk 7]
```diff
@@ -140,10 +142,13 @@ pub(crate) fn registered_tools(tcx: TyCtxt<'_>, (): ()) -> RegisteredTools {
...
         }
     }
-    // We implicitly add `rustfmt` and `clippy` to known tools,
+    // We implicitly add `rustfmt`, `clippy`, `diagnostic` to known tools,
     // but it's not an error to register them explicitly.
     let predefined_tools = [sym::clippy, sym::rustfmt];
     registered_tools.extend(predefined_tools.iter().cloned().map(Ident::with_dummy_span));
+    if tcx.features().diagnostic_namespace {
+        registered_tools.insert(Ident::with_dummy_span(sym::diagnostic));
+    }
```
Comment (estebank) [inline compiler/rustc_resolve/src/macros.rs:151]: This is ok to land on this PR as is, but I believe that as we iterate on the implementation we'll have to move the restriction of the `diagnostic` namespace away from here and onto a later stage, so that the error is not about "diagnostic not found" but rather "you're using diagnostic which is nightly only".

File: compiler/rustc_resolve/src/macros.rs [hunk 8]
```diff
@@ -500,6 +505,28 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
 
+        if self.tcx.features().diagnostic_namespace
+            && kind == MacroKind::Attr
+            && !path.segments.is_empty()
+            && path.segments[0].ident.as_str() == "diagnostic"
+        {
+            if path.segments.len() != 2 {
+                self.tcx
+                    .sess
+                    .span_err(path.span, "Diagnostic attributes requires a path with 2 segments");
```
Comment (estebank) [inline compiler/rustc_resolve/src/macros.rs:516]: We don't use title case in errors, so this would start as "diagnostic attributes". As this is referring to a name, we could go as far as using backticks around it.

File: compiler/rustc_resolve/src/build_reduced_graph.rs [hunk 9]
```diff
@@ -956,7 +956,9 @@ impl<'a, 'b, 'tcx> BuildReducedGraphVisitor<'a, 'b, 'tcx> {
                 _,
             )
             | Res::PrimTy(..)
-            | Res::ToolMod => self.r.define(parent, ident, TypeNS, (res, vis, span, expansion)),
+            | Res::ToolMod { .. } => {
```
Comment (estebank) [inline compiler/rustc_resolve/src/build_reduced_graph.rs:959]: I think this change is no longer needed either.

**Review by estebank — APPROVED** [review]

File: compiler/rustc_lint/src/context.rs [hunk 10]
```diff
@@ -404,6 +404,7 @@ impl LintStore {
             // FIXME: rustc and rustdoc are considered tools for lints, but not for attributes.
             if tool_name != sym::rustc
                 && tool_name != sym::rustdoc
+                && tool_name != sym::diagnostic
```
Comment (petrochenkov) [inline compiler/rustc_lint/src/context.rs:407]: Why is this required?
We don't have `diagnostic::foo` lints yet, only attributes.

File: compiler/rustc_resolve/src/macros.rs [hunk 11]
```diff
@@ -140,10 +142,13 @@ pub(crate) fn registered_tools(tcx: TyCtxt<'_>, (): ()) -> RegisteredTools {
             }
         }
     }
-    // We implicitly add `rustfmt` and `clippy` to known tools,
+    // We implicitly add `rustfmt`, `clippy`, `diagnostic` to known tools,
     // but it's not an error to register them explicitly.
     let predefined_tools = [sym::clippy, sym::rustfmt];
     registered_tools.extend(predefined_tools.iter().cloned().map(Ident::with_dummy_span));
+    if tcx.features().diagnostic_namespace {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:149]: Feature gates should not change any behavior, only produce errors.
`diagnostic` should always be added to registered tools, and we need to produce the feature error when it's actually used.

File: compiler/rustc_resolve/src/macros.rs [hunk 12]
```diff
@@ -500,6 +505,28 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
+                    .span_err(path.span, "Diagnostic attributes requires a path with 2 segments");
+            } else {
+                self.tcx.sess.parse_sess.buffer_lint(
+                    UNKNOWN_DIAGNOSTIC_ATTRIBUTES,
+                    path.segments[1].span(),
+                    node_id,
+                    "Unknown diagnostic attribute",
+                );
+            }
+        }
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:528]: This is not a right place for these checks, they should probably happen somewhere in `ast_validation`.

Comment (weiznich) [reply compiler/rustc_resolve/src/macros.rs:528]: Can you point to a good place for this kind of check? I'm not that well informed about the details about the compiler architecture.

File: compiler/rustc_lint/src/context.rs (hunk 10 above)
Comment (weiznich) [reply compiler/rustc_lint/src/context.rs:407]: I've just blindly added this kind of check everywhere were existing checks for the other 2 built-in tool modules existed. If that is not required here I'm happy to remove it.

File: compiler/rustc_resolve/src/macros.rs (hunk 12 above)
Comment (petrochenkov) [reply compiler/rustc_resolve/src/macros.rs:528]: `fn visit_attribute` in compiler\rustc_ast_passes\src\ast_validation.rs
It can be used for feature gating as well.

File: compiler/rustc_ast_passes/src/ast_validation.rs [hunk 13]
```diff
@@ -718,8 +719,25 @@ fn validate_generic_param_order(
...
+        {
+            let path = &attr.get_normal_item().path;
+            if path.segments.len() != 2 {
+                self.session.emit_err(InvalidDiagnosticAttribute { span: path.span });
+            } else {
+                self.lint_buffer.buffer_lint(
+                    UNKNOWN_DIAGNOSTIC_ATTRIBUTES,
+                    node_id,
+                    path.segments[1].span(),
+                    "Unknown diagnostic attribute",
```
Comment (estebank) [inline compiler/rustc_ast_passes/src/ast_validation.rs:737]:
```suggestion
                    "unknown diagnostic attribute",
```

File: compiler/rustc_ast_passes/src/ast_validation.rs [hunk 14]
```diff
@@ -718,8 +719,25 @@ fn validate_generic_param_order(
...
         validate_attr::check_attr(&self.session.parse_sess, attr);
+
+        if !attr.is_doc_comment()
+            && !attr.get_normal_item().path.segments.is_empty()
+            && attr.get_normal_item().path.segments[0].ident.name == sym::diagnostic
+        {
+            let path = &attr.get_normal_item().path;
+            if path.segments.len() != 2 {
+                self.session.emit_err(InvalidDiagnosticAttribute { span: path.span });
+            } else {
```
Comment (estebank) [inline compiler/rustc_ast_passes/src/ast_validation.rs:732]: Just to triple check, this won't trigger for a user defined `#[diagnostic]`, will it?

File: compiler/rustc_ast_passes/src/feature_gate.rs [hunk 15]
```diff
@@ -218,6 +218,14 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
                 }
             }
         }
+        if !attr.is_doc_comment()
+            && attr.get_normal_item().path.segments.len() == 2
+            && attr.get_normal_item().path.segments[0].ident.name == sym::diagnostic
+            && !self.features.diagnostic_namespace
+        {
+            let msg = "#[diagnostic] attribute name space is experimental";
```
Comment (estebank) [inline compiler/rustc_ast_passes/src/feature_gate.rs:226]:
```suggestion
            let msg = "`#[diagnostic]` attribute name space is experimental";
```

File: compiler/rustc_ast_passes/src/feature_gate.rs [hunk 16]
```diff
@@ -218,6 +218,14 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
...
+            gate_feature_post!(self, diagnostic_namespace, attr.span, msg);
```
Comment (estebank) [inline compiler/rustc_ast_passes/src/feature_gate.rs:227]: Could we use the following span instead?

```suggestion
            gate_feature_post!(
                self,
                diagnostic_namespace,
                attr.get_normal_item().path.segments[0].ident.span,
                msg
            );
```

File: compiler/rustc_ast_passes/src/ast_validation.rs (hunk 14 above)
Comment (weiznich) [reply compiler/rustc_ast_passes/src/ast_validation.rs:732]: I've just checked that and it triggers for existing user defined attributes with the name `#[diagnostic]`. I also added a test for that. That needs to be fixed, but I don't know how to best do that. Is there a way to just check here for existing attributes?

Comment (weiznich) [reply compiler/rustc_ast_passes/src/ast_validation.rs:732]: As far as I understand fixing this would require access to the name resolution results, right? In terms of compiler architecture the ast validation happens before name resolution, so these results are not there yet. At least for me that sounds like ast validation might be the wrong location for performing these kind of steps then. Maybe that means it should be moved back to the macro name resolution then? cc @petrochenkov as they've suggested to move it to ast validation. Maybe they can explain the reasoning and suggest a solution?

Comment (petrochenkov) [reply compiler/rustc_ast_passes/src/ast_validation.rs:732]: Changing the condition to `path.segments.len() > 2` would be enough.

If `#[diagnostic]` is found in AST after macro expansion, then it means that
  - either it was successfully resolved and refers to some other non-macro attribute (e.g. derive helper), and should not be reported
  - or the error "expected attribute, found tool module" was already reported in resolve, and it also should not be reported again

File: tests/ui/diagnostic_namespace/feature-gate-diagnostic_namespace.rs [hunk 17]
```diff
@@ -0,0 +1,14 @@
+// check-fail
```
Comment (petrochenkov) [inline tests/ui/diagnostic_namespace/feature-gate-diagnostic_namespace.rs:1]: `check-fail` is the default, it's not necessary to specify it explicitly.

File: compiler/rustc_ast_passes/messages.ftl [hunk 18]
```diff
@@ -159,6 +159,8 @@ ast_passes_inherent_cannot_be = inherent impls cannot be {$annotation}
     .type = inherent impl for this type
     .only_trait = only trait implementations may be annotated with {$annotation}
 
+ast_passes_invalid_diagnostic_attribute = attributes starting with `diagnostic` requires a path with 2 segments
```
Comment (petrochenkov) [inline compiler/rustc_ast_passes/messages.ftl:162]:
```suggestion
ast_passes_invalid_diagnostic_attribute = attributes starting with `diagnostic` require a path with 2 segments
```

**Comment by petrochenkov** [comment]
> `fn visit_attribute` in compiler\rustc_ast_passes\src\ast_validation.rs It can be used for feature gating as well.

Ok, now I see the result and I don't like it.
Let's move all that stuff (both feature gate and linting) back to `fn smart_resolve_macro_path`, my apologies.
@rustbot author

**Comment by weiznich** [comment]
> Let's move all that stuff (both feature gate and linting) back to fn smart_resolve_macro_path, my apologies.

I've removed the relevant changes for the linting from the PR, but I left the feature gating in place as it's just a local addition to the feature gate pass and it seems to fit there quite well. If that's wrong I will happily move it back as well, but I assume the main problem with the other change was the addition of the `NodeId` parameter to the relevant ast-pass function. As that's not required for the feature gate it might be fine to keep it there?

@rustbot reviewer

File: tests/ui/diagnostic_namespace/requires_path.stderr [hunk 19]
```diff
@@ -0,0 +1,8 @@
+error: cannot find attribute `diagnostic` in this scope
+  --> $DIR/requires_path.rs:3:3
+   |
+LL | #[diagnostic]
+   |   ^^^^^^^^^^
```
Comment (estebank) [inline tests/ui/diagnostic_namespace/requires_path.stderr:5]: Before stabilization of the namespace, we'll want to customize this error to be something along the lines of "`#[diagnostic]` needs to be followed by one of `*`".

File: compiler/rustc_resolve/src/macros.rs [hunk 20]
```diff
@@ -500,6 +502,18 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             Err(Determinacy::Undetermined) => return Err(Indeterminate),
         };
 
+        if kind == MacroKind::Attr
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:505]:
```suggestion
        if res == Res::NonMacroAttr(NonMacroAttrKind::Tool)
```
This is incorrect because we can encounter macro attribute resolutions here.

This logic should also be run at the very end of `smart_resolve_macro_path`, after all the "unexpected resolution" errors are reported.

File: compiler/rustc_resolve/src/macros.rs [hunk 21]
```diff
@@ -500,6 +502,18 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
+            && path.segments.len() == 2
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/macros.rs:506]:
```suggestion
            && path.segments.len() >= 2
```
Looks like `#[diagnostic::foo::bar]` is not reported in any way, but it probably should be.

Technically this condition can be removed entirely because tool attributes are always >= 2 segments.

File: compiler/rustc_resolve/src/macros.rs (hunk 20 above)
Comment (petrochenkov) [reply compiler/rustc_resolve/src/macros.rs]: The second part of the comment wasn't addressed.

**Comment by petrochenkov** [comment]
r=me after addressing https://github.com/rust-lang/rust/pull/111780#discussion_r1277359966 and squashing commits.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+
